Fixes Issue2099: enforce minimum `tx` fees

hewison-chris opened the pull request:
Adding check to transactions when there is no data payload to also have a minimum fee available.

Commit:
Fixed minimum fee check for empty payload

Previously the minimum fee was only checked if the payload had a data
length > 0

hewison-chris committed:
Use fee rate per byte in TxBuilder

This enables the setting of a fee per byte rate for calculating the tx
fees in TxBuilder. If there is a very small refund then it is not added
as an output but implicitly included in the fees, where fees are the sum
of the input values minus the sum of the output values.

Geod24 commented on source/agora/utils/TxBuilder.d:
```diff
@@ -299,7 +299,7 @@ public struct TxBuilder
         in OutputType outputs_type = OutputType.Payment, uint unlock_age = 0,
         Unlock delegate (in Transaction tx, in OutputRef out_ref) @safe nothrow
-        unlocker = null) @safe nothrow
+        unlocker = null, in Amount fees = Amount(10_000)) @safe nothrow
```
Let's make this a member of the struct that can be `set` instead of yet another `sign` parameter.

hewison-chris replied:
Will make setter for this.

Geod24 commented on source/agora/utils/TxBuilder.d:
```diff
@@ -319,20 +319,22 @@ public struct TxBuilder
         foreach (ref o; this.data.outputs)
             o.type = outputs_type;
 
+        // Sign all inputs using unlocker
+        foreach (idx, ref in_; this.inputs)
+            this.data.inputs[idx].unlock = unlocker(this.data, in_);
+
```
That will not work, the signature needs to happen after we have all the `output` ready, otherwise it will be invalid.

hewison-chris replied:
At first I thought this was only `inputs` being updated so could be done first, but you are right as the `unlocker` will sign the `tx` which needs all the outputs to be included.

Geod24 commented on source/agora/utils/TxBuilder.d:
```diff
@@ -319,20 +319,22 @@ public struct TxBuilder
+
         // Add the refund output if needed
-        if (this.leftover.value > Amount(0))
+        if (this.leftover.value > fees)
```
`fee` should not be an absolute value but a ration per byte, and this function needs to compute the correct size.

hewison-chris replied:
Yes as a first step I thought we could set fees high enough to prevent all txs being rejected. Then introduce a calculated fee based on the size of the tx.

Geod24 replied:
I think going straight for the calculated fee is better. A set fee is just an extra output with some specific value, we can already do that manually. Plus, we won't have to change the tests twice.

hewison-chris replied:
There is no output for the fee. Just the leftover from total input values - total output values. I'm thinking we maybe  don't need to calculate in TxBuilder as it is just generating test transactions. The calculated fee per byte happens when we validate and decide on candidates.

hewison-chris commented:
The fee rate is now included in `TxBuilder` and a fee calculated based on the `tx` byte size. 
However, this is still a WIP as the `TxBuilder` tests will certainly fail until rewritten. 
Also, I expect quite a few other tests will also need to be updated. 
I will work together with @kchulj on Monday to continue fixing up the tests. :-)

hewison-chris commented on source/agora/test/Script.d:
```diff
@@ -135,21 +135,21 @@ unittest
     assert(blocks.length == 1);
     sort(txs_0);
-    assert(blocks[0].txs == txs_0);
+    assert(blocks[0].txs.filter!(tx => !tx.isCoinbase).array == txs_0);
```
We agreed we will not send `Coinbase` txs every block so these need to be removed.

hewison-chris replied:
Updated

hewison-chris reviewed (commented):
Need to check why we get `Coinbase tx` per block with fees added.

Geod24 commented on source/agora/consensus/state/UTXODB.d:
```diff
@@ -73,7 +73,7 @@ package class UTXODB
     {
         return () @trusted {
-            auto results = this.db.execute("SELECT hash, unlock_height, type, amount, locktype, lock FROM utxo");
+            auto results = this.db.execute("SELECT hash, unlock_height, type, amount, locktype, lock FROM utxo order by hash");
```
Use uppercase for SQL keywords, as done everywhere else.
Also that's a good idea, could you make it its own PR ?

hewison-chris replied:
Updated in new PR https://github.com/bosagora/agora/pull/2348

Geod24 commented on source/agora/node/Validator.d:
```diff
@@ -606,8 +606,8 @@ public class Validator : FullNode, API
         const enrollment = this.enroll_man.createEnrollment(enroll_key, avail_height);
-        log.trace("Sending Enrollment for enrolling at height {} (to validate blocks {} to {})",
-            avail_height, avail_height + 1, avail_height + this.params.ValidatorCycle);
+        log.info("Sending Enrollment with utxo {} for enrolling at height {} (to validate blocks {} to {})",
```
The reason it is at trace level currently is because it happens multiple times, not just once.
Essentially every block from X block before the enrollment runs out (where X is configured by the node).

hewison-chris replied:
Do you think once per block is too much for an important event like sending an enrollment?

hewison-chris replied:
In fact this is not relevant to this PR anymore so will remove.

hewison-chris commented:
> Need to check why we get `Coinbase tx` per block with fees added.

This is now fixed by only creating Coinbase at payout period (default is every 144 blocks)

hewison-chris commented on source/agora/utils/TxBuilder.d:
```diff
@@ -493,31 +559,44 @@ public struct TxBuilder
+    auto refund = total - sum;
+    assert(tx.outputs.count!(o => o.value == refund) == 1);
 }
 
```
The rest of the tests in this module need to be updated similar to the one above.

kchulj replied:
Will update the tests on a local branch.

kchulj commented:
> // This transaction has 16 txs

> // This transaction has 4 txs (3 targets + 1 refund)

Some of these comments are inaccurate (should be outputs).
I will fix this in a separate PR.

kchulj commented:
Some of the tests assume that there is only one output (no refund output).
What should we do in this case? It seems like the TxBuilder methods assume that there is always a refund output.
`total_fees.mul(this.data.sizeInBytes + refund_byte_size)`

hewison-chris commented:
> Some of the tests assume that there is only one output (no refund output).
> What should we do in this case? It seems like the TxBuilder methods assume that there is always a refund output.
> `total_fees.mul(this.data.sizeInBytes + refund_byte_size)`

Yes we will always assume there is a refund to make sure if we do end up with one that we have enough fees included. I think we should have tests that cover both with and without a refund.

hewison-chris commented:
This now includes the rest of the TxBuilder tests fixed thanks to @kchulj . 🏅

Geod24 commented on source/agora/utils/TxBuilder.d:
```diff
@@ -155,10 +155,12 @@ public struct TxBuilder
     public ref typeof(this) attach (const Transaction tx)
         @safe nothrow return
     {
+        // Don't include small change outputs less than required fees
         this.inputs ~= iota(tx.outputs.length)
             .map!(index => OutputRef(tx.outputs[index], Input(tx, index).utxo))
+            .filter!(o => o.output.value > Amount(5000))
```
Why ? That seems rather arbitrary. I think the caller should be at liberty to do this. It's obviously wasteful to do so, but the job of `TxBuilder` is to build a transaction, not to define best practices for doing so.
So unless this would systematically lead to invalid transactions being created, I don't see a reason to do it.
Moreover, doing it *silently* would lead to hard to debug issues, where a user wrongly uses `TxBuilder` and gets no feedback (imagine the user having only <= 5k amounts outputs, trying to use TxBuilder, and seeing it empty).

hewison-chris replied:
I was forgetting that you can attach multiple times to end up with enough in leftover to cover the fees. So this will be reverted and the calling code will have to handle the cases where the output is small change and would create invalid transactions.

linked0 commented on source/agora/test/Script.d:
```diff
@@ -74,11 +74,11 @@ unittest
     const lock_height_2 = Height(2);
     auto unlock_height_2 = block_6.spendable()
-        .map!(txb => txb.lock(lock_height_2).sign(OutputType.Payment, 0, &keyUnlocker));
+        .map!(txb => txb.unlock_signer(&keyUnlocker).lock(lock_height_2).sign(OutputType.Payment, 0));
```
Out of curiosity, what's the reason this introduces the new function, `unlock_signer`?

hewison-chris replied:
We want to calculate the fees when doing `split` so need to know the `unlocker` so that the size of the `input`s can be calculated. Previously it was passed as a param to `sign`.

Geod24 commented on source/agora/common/Amount.d:
```diff
@@ -139,6 +139,34 @@ public struct Amount
         return copy;
     }
 
+    /// Ditto
+    public Amount opBinary (string op : "*") (in ulong multiplier) const scope
+        @safe
+    {
+        Amount copy = this;
+        if (!copy.mul(multiplier))
+            throw new Exception(format("Amount %d cannot be multiplied by %d", this, multiplier));
+        return copy;
+    }
+
+    /// Return the amount divided by denominator ignoring the remainder (use % to get remainder)
+    public Amount opBinary (string op : "/") (in ulong denominator) const scope
+        @safe
+    {
+        Amount copy = this;
+        copy.div(denominator);
+        return copy;
+    }
+
+    /// Return the remainder of division
+    public Amount opBinary (string op : "%") (in ulong denominator) const scope
+        @safe
+    {
+        Amount copy = this;
+        return copy.div(denominator);
+    }
+
+
```
The mul makes sense, but the div / modulo, where is it used ? The core tenet of `Amount` is that you can't get it wrong. The `opBinary` additions still respect this, by giving your a runtime Exception (having just `add` / `sub` meant you were forced to handle it statically). But discarding part of the `Amount` is not something we should encourage, so if there's only one or two usages of it, I'd rather not introduce the overloads.

hewison-chris replied:
How about if we make them `version unittest`. The idea was to make the test code less verbose and easier to understand.

Geod24 replied:
How often are they used though ? I don't expect it to be that often ?

hewison-chris replied:
I refactored to not need the `/` and `%` `opBinary` functions as we can not make it `version unittest` as it is used in other tests which are not `version unittest` like block storage test.

Geod24 commented on source/agora/utils/TxBuilder.d:
```diff
@@ -471,161 +531,240 @@ public struct TxBuilder
-        this.leftover.value = Amount(0);
-
-        return this;
+        return this.leftover.value >= minFees(true);
```
`this.`

hewison-chris replied:
Updated

Geod24 commented on source/agora/utils/TxBuilder.d:
```diff
@@ -471,161 +531,240 @@ public struct TxBuilder
-    private Output leftover;
+    private Amount minFees (bool include_refund_fees = false) nothrow @safe
+    {
+        auto unlocker = this.unlocker == null ? &keyUnlocker : this.unlocker;
```
That's a good example of a special case that can be easily avoided.

hewison-chris replied:
Not so easy but have found a solution.

hewison-chris commented:
It looks like there are still quite a few updates needed to support minimum fees in the tests for Flash.

hewison-chris commented:
> It looks like there are still quite a few updates needed to support minimum fees in the tests for Flash.

For now I have set min fee rate to zero for these tests so that other tests can be fixed first.

Geod24 commented on source/agora/consensus/Fee.d:
```diff
@@ -238,7 +238,7 @@ public class FeeManager
             return "Fee: Transaction size overflows fee cap";
         if (tx_fee < minimumFee)
-            return "Transaction: Fee is less than adjusted minimum fee";
+            return "Transaction: Insufficient Implicit Fees";
```
"Insufficient fee for transaction size"

I get what you mean by "implicit" fee (fees that are not explicit, which are those of the payload), but it is the main fee - what users will see the most. Hence qualifying it might be confusing for users, but to respect the goal of differentiating it from the payload fee, "for transaction" should do, what do you think ?

Geod24 replied:
BTW this is in the second commit, it shouldn't. In fact you could submit it as a PR to HEAD in its own commit if you want.

hewison-chris replied:
I'll extract it into new PR and reword.

hewison-chris replied:
> There's a lot of mixed up changes. It would be best to split the various fixes in their own commit / PR. And we probably will need to have the `TxBuilder` part merged before we actually enable the checks in consensus, so that Faucet starts to create valid txs, right ?

Yes I will split up the changes so we can get Faucet updated to be compatible first.

Geod24 commented on source/agora/node/Ledger.d:
```diff
@@ -583,6 +583,8 @@ public class Ledger
         in uint[] missing_validators) nothrow @safe
     {
+        if (height % this.params.PayoutPeriod != 0)    // This is not Coinbase payout block
+            return [];
```
`null`

hewison-chris replied:
Updated

Geod24 commented on source/agora/test/Base.d:
```diff
@@ -1984,7 +1984,7 @@ public struct TestConf
         /// The minimum (transaction size adjusted) fee.
         /// Transaction size adjusted fee = tx fee / tx size in bytes.
-        min_fee: Amount(0),
+        // min_fee: Amount(700),
```
Just remove it ?

hewison-chris replied:
Removed

Geod24 commented on source/agora/utils/TxBuilder.d:
```diff
@@ -227,9 +231,52 @@ public struct TxBuilder
+        return Unlock([ubyte(64)] ~ sig.toBlob()[]);
+    }
+
+    public static Unlock signWithSpecificKeyTrue (KeyPair key) (in Transaction tx, in OutputRef out_ref)
+        @safe nothrow
+    {
+        auto sig = key.sign(tx);
+        return Unlock([ubyte(64)] ~ sig.toBlob()[] ~ [ubyte(OP.TRUE)]);
+    }
+
+    public static Unlock signWithSpecificKeyFalse (KeyPair key) (in Transaction tx, in OutputRef out_ref)
+        @safe nothrow
+    {
+        auto sig = key.sign(tx);
+        return Unlock([ubyte(64)] ~ sig.toBlob()[] ~ [ubyte(OP.FALSE)]);
+    }
```
What are those ?

hewison-chris replied:
These are used in `Flash` tests. Shall I make it template function with `opcode` `ubyte`, or do you prefer I keep those functions in the `Flash` unit test?

Geod24 replied:
Well the `true` / `false` could just be a runtime parameter, no ?

hewison-chris replied:
This is the function passed to the `txBuilder` for signing so I think the extra parameter would not be compatible.

Geod24 commented:
There's a lot of mixed up changes. It would be best to split the various fixes in their own commit / PR. And we probably will need to have the `TxBuilder` part merged before we actually enable the checks in consensus, so that Faucet starts to create valid txs, right ?

hewison-chris committed:
Coinbase transactions only at each `payoutPeriod`

`Coinbase outputs` are only created every `ConsensusParams.PayoutPeriod`
number of blocks. Previously only Coinbase transactions for the Commons
Budget funding were using this interval, which meant that when adding
fees to transactions we would create Coinbase transactions every block.

hewison-chris committed:
Unit test ValidatorRecurringEnrollment fix

The test was not really testing enrolling when coming back on line as
the nodes were still enrolling whilst they were in sleep mode. To fix
this the nodes are also made to skip enrolling before being woken up.

hewison-chris committed:
Only auto enroll near the end of the cycle

When a validator is configured to auto enroll it should send enrollments
when near the end of the cycle rather then every block.

hewison-chris committed:
Unit test comments: ValidatorRecurringEnrollemnt

hewison-chris committed:
Unit test update: ValidatorRecurringEnrollment

When node #4 is catching up we don't need to check the first 4 nodes.

hewison-chris committed:
Unit test update: ValidatorRecurringEnrollment

Updated node #5 catch up to use `expectHeightAndPreImg` rather than
`expectHeight` which is more appropriate as it makes sure `pre-images`
are available before `this.setTimeFor(height)` is called.

hewison-chris commented:
Closing this PR as changes are extracted into other branches to be merged.
